src/core/common/message.cpp

```diff
@@ -294,12 +294,34 @@ ThreadError Message::Prepend(const void *aBuf, uint16_t aLength)
     mInfo.mLength += aLength;
     SetOffset(GetOffset() + aLength);
 
-    Write(0, aLength, aBuf);
+    if (aBuf != NULL)
+    {
+        Write(0, aLength, aBuf);
+    }
 
 exit:
     return error;
 }
 
+ThreadError Message::RemoveHeader(uint16_t aLength)
```

Comment:
I agree that using `Message::RemoveHeader()` in this case makes sense.  Logically, after the outer IP header is removed the inner IP datagram should run through `Ip6::HandleDatagram()` again.